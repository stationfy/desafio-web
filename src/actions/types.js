export const FETCH_REPOSITORIES = "FETCH_REPOSITORIES";
export const FETCH_REPOSITORIES_SUCCESS = "FETCH_REPOSITORIES_SUCCESS";
export const FETCH_REPOSITORIES_FAIL = "FETCH_REPOSITORIES_FAIL"

export const FETCH_PULL_REQUESTS = "FETCH_PULL_REQUESTS";
export const FETCH_PULL_REQUESTS_SUCCESS = "FETCH_PULL_REQUESTS_SUCCESS";
export const FETCH_PULL_REQUESTS_FAIL = "FETCH_PULL_REQUESTS_FAIL";