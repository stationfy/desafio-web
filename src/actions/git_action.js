import axios from "axios";
import {
  FETCH_REPOSITORIES,
  FETCH_REPOSITORIES_SUCCESS,
  FETCH_REPOSITORIES_FAIL,
  FETCH_PULL_REQUESTS,
  FETCH_PULL_REQUESTS_SUCCESS,
  FETCH_PULL_REQUESTS_FAIL
} from "./types";

const ROOT_URL = "https://api.github.com/";

export function fetchRepositories(page) {
  return function(dispatch) {
    dispatch({
      type: FETCH_REPOSITORIES
    });

    const url = `${ROOT_URL}search/repositories?q=language:Javascript&sort=stars&page=${page}`;
    axios
      .get(url)
      .then(res => {
        dispatch({
          type: FETCH_REPOSITORIES_SUCCESS,
          payload: res
        });
      })
      .catch(error => {
        dispatch({
          type: FETCH_REPOSITORIES_FAIL,
          payload: error.message
        });
      });
  };
}

export function fetchPullRequests(creator, repo) {
  return function(dispatch) {
    dispatch({
      type: FETCH_PULL_REQUESTS
    });

    const url = `${ROOT_URL}repos/${creator}/${repo}/pulls`;
    axios
      .get(url)
      .then(res => {
        dispatch({
          type: FETCH_PULL_REQUESTS_SUCCESS,
          payload: res
        });
      })
      .catch(error => {
        dispatch({
          type: FETCH_PULL_REQUESTS_FAIL,
          payload: error.message
        });
      });
  };
}
