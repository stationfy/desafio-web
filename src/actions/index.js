import {CLICK} from '../constants';

export const click = (project, owner, pulls) => {
    return{
      type: CLICK,
      project,
      owner,
      pulls,
    }
}
