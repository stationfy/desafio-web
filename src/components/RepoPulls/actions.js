import * as c from "./constants";

export const fetchPulls = ({ creator, repo }) => (
  dispatch
) => {
  dispatch({
    type: c.FETCH_PULLS,
    payload: {
      request: {
        url: `/repos/${creator}/${repo}/pulls`,
        method: "GET"
      },
      options: {
        onSuccess: ({ getState, dispatch, response }) => {
          dispatch({
            type: c.FETCH_PULLS_SUCCESS,
            payload: {
              data: response.data
            }
          });
        },
        onError: ({ getState, dispatch, error }) => {
          dispatch({
            type: c.FETCH_PULLS_FAIL,
            payload: {
              error
            }
          })
        }
      }
    }
  })
};