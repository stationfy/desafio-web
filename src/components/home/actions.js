import * as c from "./constants";

export const fetchRepos = (payload) => (
  dispatch
) => {
  const { queryString, sort, page, order } = payload;
  return dispatch({
    type: c.FETCH_REPOS,
    payload: {
      request: {
        url: `/search/repositories?q=${queryString}&sort=${sort}&page=${page}&order=${order}`,
        method: "GET"
      },
      options: {
        onSuccess: ({ getState, dispatch, response }) => {
          const state = getState();
          dispatch({
            type: c.FETCH_REPOS_SUCCESS,
            payload: {
              data: {...response.data, page: state.home.page + 1 }
            }
          });
        },
        onError: ({ getState, dispatch, error }) => {
          dispatch({
            type: c.FETCH_REPOS_FAIL,
            payload: {
              error
            }
          })
        }
      }
    }
  })
}