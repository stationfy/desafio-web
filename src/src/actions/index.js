import {
    REQUEST_REPOS,
    RECEIVE_REPOS,
    RECEIVE_REPOS_ERROR,
    REQUEST_PULLS,
    RECEIVE_PULLS,
    RECEIVE_PULLS_ERROR,
    RECEIVE_REPOS_PAGE_COUNT,
    RECEIVE_PULLS_PAGE_COUNT,
    SET_SCROLLING,
    CLEAR_PULLS
} from '../actionTypes.json'

export function requestRepos() {
    return { type: REQUEST_REPOS }
}

function receiveRepos(repos) {
    return { type: RECEIVE_REPOS, repos }
}

function receiveReposErr(error) {
    return { type: RECEIVE_REPOS_ERROR, error }
}

export function requestPulls() {
    return { type: REQUEST_PULLS }
}

function receivePulls(pulls, full_name) {
    return { type: RECEIVE_PULLS, pulls, full_name }
}

function receivePullsErr(error, full_name) {
    return { type: RECEIVE_PULLS_ERROR, error, full_name }
}

function receiveRepoPageCount(count) {
    return { type: RECEIVE_REPOS_PAGE_COUNT, count }
}

function receivePullsPageCount(count) {
    return { type: RECEIVE_PULLS_PAGE_COUNT, count }
}

function setScrolling(isScrolling) {
    return { type: SET_SCROLLING, isScrolling }
}

function clearPulls(getting) {
    return { type: CLEAR_PULLS, getting }
}

export function fetchRepos(page) {
    return async dispatch => {
        dispatch(requestRepos())
        try {
            const res = await fetch(`https://api.github.com/search/repositories?q=language:Javascript&sort=stars&page=${page}`)
            const { items = [] } = await res.json()
            const repos = items.length ? items.map(repo => {
                const {
                    id,
                    name,
                    description,
                    forks,
                    stargazers_count,
                    owner: { avatar_url: ownerAvatar, login: ownerName  },
                    full_name
                } = repo
    
                return { 
                    id,
                    name,
                    description: `${description.slice(0, 67)}...`,
                    forks,
                    stars: stargazers_count,
                    ownerAvatar,
                    ownerName,
                    full_name
                }
            }) : []

            return dispatch(receiveRepos(repos))
        }
        catch (err) {
            return dispatch(receiveReposErr(err))
        }
    }
}

export function fetchPulls(full_name, page) {
    return async dispatch => {
        dispatch(requestPulls())
        try {
            const res = await fetch(`https://api.github.com/repos/${full_name}/pulls?page=${page}`)
            const json = await res.json()
            const pulls = json.length ? json.map(pull => {
                const {
                    id,
                    title,
                    body,
                    created_at,
                    user: { avatar_url, login },
                    html_url: link
                } = pull
    
                return {
                    id,
                    title: `${title ? title.slice(0, 47) : 'Pull without title'}...`,
                    body: `${body ? body.slice(0, 67) : 'Pull without body'}...`,
                    createdAt: new Date(created_at).toLocaleDateString(),
                    authorAvatar: avatar_url,
                    authorLogin: login,
                    link
                }
            }) : []

            return dispatch(receivePulls(pulls, full_name))
        }
        catch (err) {
            return dispatch(receivePullsErr(err))
        }
    }
}

export function setReposPageCount(count) {
    return dispatch => dispatch(receiveRepoPageCount(count))
}

export function setPullsPageCount(count) {
    return dispatch => dispatch(receivePullsPageCount(count))
}

export function scrolling(isScrolling) {
    return dispatch => dispatch(setScrolling(isScrolling))
}

export function clearPullsState(getting) {
    return dispatch => dispatch(clearPulls(getting))
}