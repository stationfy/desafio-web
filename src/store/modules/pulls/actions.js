export const fetchPullsRequest = payload => ({
  type: "@pulls/FETCH_PULLS_REQUEST",
  payload
});

export const fetchPullsRequestSuccess = payload => ({
  type: "@pulls/FETCH_PULLS_REQUEST_SUCCESS",
  payload
});
