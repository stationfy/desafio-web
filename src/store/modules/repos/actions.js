export const fetchReposRequest = payload => ({
  type: "@repos/FETCH_REPOS_REQUEST",
  payload
});

export const fetchReposRequestSuccess = payload => ({
  type: "@repos/FETCH_REPOS_REQUEST_SUCCESS",
  payload
});
