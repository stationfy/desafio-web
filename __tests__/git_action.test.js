import configureMockStore from "redux-mock-store";
import moxios from "moxios";
import expect from "expect";
import thunk from "redux-thunk";
import { fetchRepositories, fetchPullRequests }  from "../src/actions/git_action";
import { FETCH_REPOSITORIES_SUCCESS } from "../src/actions/types";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe("get Repositories from page 1", () => {
  it("creates FETCH_REPOSITORIES_SUCCESS after successfuly fetching response", () => {
    expect(fetchRepositories(1)).toMatchSnapshot();
  });
});

describe("get pull requests from creator and repo", () => {
  it('create FETCH_PULL_REQUESTS_SUCCESS after successfuly fetching response', () => {
    expect(fetchPullRequests('facebook', 'react')).toMatchSnapshot();
  })
});