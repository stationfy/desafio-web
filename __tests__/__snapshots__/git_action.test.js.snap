// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get Repositories from page 1 creates FETCH_REPOSITORIES_SUCCESS after successfuly fetching response 1`] = `[Function]`;

exports[`get pull requests from creator and repo create FETCH_PULL_REQUESTS_SUCCESS after successfuly fetching response 1`] = `[Function]`;
